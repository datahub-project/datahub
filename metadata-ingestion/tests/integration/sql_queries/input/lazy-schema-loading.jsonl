{"query": "SELECT user_id, session_id FROM snowflake.db.sessions WHERE created_at > '2021-01-01'", "timestamp": 1609459200, "user": "etl_job", "operation_type": "SELECT"}
{"query": "INSERT INTO snowflake.db.audit_log SELECT user_id, 'login' as action, NOW() as timestamp FROM snowflake.db.sessions WHERE user_id IS NOT NULL", "timestamp": 1609459260, "user": "etl_job", "operation_type": "INSERT"}
{"query": "CREATE TEMPORARY TABLE temp_user_sessions AS SELECT user_id, COUNT(*) as session_count FROM snowflake.db.sessions GROUP BY user_id", "timestamp": 1609459320, "user": "analyst", "operation_type": "CREATE"}
{"query": "SELECT * FROM temp_user_sessions WHERE session_count > 5", "timestamp": 1609459380, "user": "analyst", "operation_type": "SELECT"}
{"query": "DROP TABLE temp_user_sessions", "timestamp": 1609459440, "user": "analyst", "operation_type": "DROP"}

