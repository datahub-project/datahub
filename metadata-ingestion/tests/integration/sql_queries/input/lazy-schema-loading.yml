source:
  type: sql-queries
  config:
    query_file: ./input/lazy-schema-loading.jsonl
    platform: snowflake
    use_schema_resolver: true

sink:
  type: file
  config:
    filename: ./output.json

datahub_api:
  server: http://localhost:8080
