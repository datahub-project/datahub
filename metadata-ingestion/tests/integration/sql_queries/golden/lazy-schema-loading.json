[
{
    "entityType": "dataset",
    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD)",
    "changeType": "UPSERT",
    "aspectName": "upstreamLineage",
    "aspect": {
        "json": {
            "upstreams": [
                {
                    "auditStamp": {
                        "time": 1757897989859,
                        "actor": "urn:li:corpuser:_ingestion"
                    },
                    "created": {
                        "time": 1609459260000,
                        "actor": "urn:li:corpuser:etl_job"
                    },
                    "dataset": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD)",
                    "type": "TRANSFORMED",
                    "query": "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9"
                }
            ],
            "fineGrainedLineages": [
                {
                    "upstreamType": "FIELD_SET",
                    "upstreams": [
                        "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD),user_id)"
                    ],
                    "downstreamType": "FIELD",
                    "downstreams": [
                        "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD),user_id)"
                    ],
                    "transformOperation": "COPY: \"SESSIONS\".\"USER_ID\" AS \"USER_ID\"",
                    "confidenceScore": 0.2,
                    "query": "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989860,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9",
    "changeType": "UPSERT",
    "aspectName": "queryProperties",
    "aspect": {
        "json": {
            "customProperties": {},
            "statement": {
                "value": "INSERT INTO snowflake.db.audit_log SELECT user_id, 'login' as action, NOW() as timestamp FROM snowflake.db.sessions WHERE user_id IS NOT NULL",
                "language": "SQL"
            },
            "source": "SYSTEM",
            "created": {
                "time": 1609459260000,
                "actor": "urn:li:corpuser:etl_job"
            },
            "lastModified": {
                "time": 1609459260000,
                "actor": "urn:li:corpuser:etl_job"
            }
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989860,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9",
    "changeType": "UPSERT",
    "aspectName": "querySubjects",
    "aspect": {
        "json": {
            "subjects": [
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD),user_id)"
                },
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD),user_id)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD),action)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD),timestamp)"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989861,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9",
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
        "json": {
            "platform": "urn:li:dataPlatform:snowflake"
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989861,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "dataset",
    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD)",
    "changeType": "UPSERT",
    "aspectName": "datasetUsageStatistics",
    "aspect": {
        "json": {
            "timestampMillis": 1609459200000,
            "eventGranularity": {
                "unit": "DAY",
                "multiple": 1
            },
            "partitionSpec": {
                "partition": "FULL_TABLE_SNAPSHOT",
                "type": "FULL_TABLE"
            },
            "uniqueUserCount": 2,
            "totalSqlQueries": 3,
            "topSqlQueries": [
                "SELECT user_id, session_id FROM snowflake.db.sessions WHERE created_at > '2021-01-01'",
                "INSERT INTO snowflake.db.audit_log SELECT user_id, 'login' as action, NOW() as timestamp FROM snowflake.db.sessions WHERE user_id IS NOT NULL",
                "CREATE TEMPORARY TABLE temp_user_sessions AS SELECT user_id, COUNT(*) as session_count FROM snowflake.db.sessions GROUP BY user_id"
            ],
            "userCounts": [
                {
                    "user": "urn:li:corpuser:etl_job",
                    "count": 2
                },
                {
                    "user": "urn:li:corpuser:analyst",
                    "count": 1
                }
            ],
            "fieldCounts": [
                {
                    "fieldPath": "user_id",
                    "count": 3
                },
                {
                    "fieldPath": "session_id",
                    "count": 1
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989862,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "dataset",
    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD)",
    "changeType": "UPSERT",
    "aspectName": "datasetUsageStatistics",
    "aspect": {
        "json": {
            "timestampMillis": 1609459200000,
            "eventGranularity": {
                "unit": "DAY",
                "multiple": 1
            },
            "partitionSpec": {
                "partition": "FULL_TABLE_SNAPSHOT",
                "type": "FULL_TABLE"
            },
            "uniqueUserCount": 1,
            "totalSqlQueries": 1,
            "topSqlQueries": [
                "SELECT * FROM temp_user_sessions WHERE session_count > 5"
            ],
            "userCounts": [
                {
                    "user": "urn:li:corpuser:analyst",
                    "count": 1
                }
            ],
            "fieldCounts": [
                {
                    "fieldPath": "session_count",
                    "count": 1
                },
                {
                    "fieldPath": "user_id",
                    "count": 1
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989863,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "dataset",
    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.audit_log,PROD)",
    "changeType": "UPSERT",
    "aspectName": "operation",
    "aspect": {
        "json": {
            "timestampMillis": 1757897989863,
            "partitionSpec": {
                "partition": "FULL_TABLE_SNAPSHOT",
                "type": "FULL_TABLE"
            },
            "actor": "urn:li:corpuser:etl_job",
            "operationType": "INSERT",
            "sourceType": "DATA_PLATFORM",
            "lastUpdatedTimestamp": 1609459260000,
            "queries": [
                "urn:li:query:ac1bdb4a206870224b316a0412c4b4b57544f56f9e8618d490546d8268c26ab9"
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989863,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "dataset",
    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD)",
    "changeType": "UPSERT",
    "aspectName": "operation",
    "aspect": {
        "json": {
            "timestampMillis": 1757897989864,
            "partitionSpec": {
                "partition": "FULL_TABLE_SNAPSHOT",
                "type": "FULL_TABLE"
            },
            "actor": "urn:li:corpuser:analyst",
            "operationType": "UNKNOWN",
            "sourceType": "DATA_PLATFORM",
            "lastUpdatedTimestamp": 1609459440000,
            "queries": [
                "urn:li:query:bcc12e17f79999d69b8af00cd9763d210c81b2395b4c9011a08111d5dc47ae6f"
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989864,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:bcc12e17f79999d69b8af00cd9763d210c81b2395b4c9011a08111d5dc47ae6f",
    "changeType": "UPSERT",
    "aspectName": "queryProperties",
    "aspect": {
        "json": {
            "customProperties": {},
            "statement": {
                "value": "DROP TABLE temp_user_sessions",
                "language": "SQL"
            },
            "source": "SYSTEM",
            "created": {
                "time": 1609459440000,
                "actor": "urn:li:corpuser:analyst"
            },
            "lastModified": {
                "time": 1609459440000,
                "actor": "urn:li:corpuser:analyst"
            }
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989865,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:bcc12e17f79999d69b8af00cd9763d210c81b2395b4c9011a08111d5dc47ae6f",
    "changeType": "UPSERT",
    "aspectName": "querySubjects",
    "aspect": {
        "json": {
            "subjects": [
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD)"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989865,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:bcc12e17f79999d69b8af00cd9763d210c81b2395b4c9011a08111d5dc47ae6f",
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
        "json": {
            "platform": "urn:li:dataPlatform:snowflake"
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989865,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:abd3fa695178dd5991b4d94128c35cf629ade4be2ac6952211f12b3ee005ec5b",
    "changeType": "UPSERT",
    "aspectName": "queryProperties",
    "aspect": {
        "json": {
            "customProperties": {},
            "statement": {
                "value": "SELECT user_id, session_id FROM snowflake.db.sessions WHERE created_at > '2021-01-01'",
                "language": "SQL"
            },
            "source": "SYSTEM",
            "created": {
                "time": 1609459200000,
                "actor": "urn:li:corpuser:etl_job"
            },
            "lastModified": {
                "time": 1609459200000,
                "actor": "urn:li:corpuser:etl_job"
            }
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989866,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:abd3fa695178dd5991b4d94128c35cf629ade4be2ac6952211f12b3ee005ec5b",
    "changeType": "UPSERT",
    "aspectName": "querySubjects",
    "aspect": {
        "json": {
            "subjects": [
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD),session_id)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD),user_id)"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989866,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:abd3fa695178dd5991b4d94128c35cf629ade4be2ac6952211f12b3ee005ec5b",
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
        "json": {
            "platform": "urn:li:dataPlatform:snowflake"
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989867,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:d2a535662522728348ecbfc35b114fa4d4ec74c797735a3126d6d53e279fa452",
    "changeType": "UPSERT",
    "aspectName": "queryProperties",
    "aspect": {
        "json": {
            "customProperties": {},
            "statement": {
                "value": "CREATE TEMPORARY TABLE temp_user_sessions AS SELECT user_id, COUNT(*) as session_count FROM snowflake.db.sessions GROUP BY user_id",
                "language": "SQL"
            },
            "source": "SYSTEM",
            "created": {
                "time": 1609459320000,
                "actor": "urn:li:corpuser:analyst"
            },
            "lastModified": {
                "time": 1609459320000,
                "actor": "urn:li:corpuser:analyst"
            }
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989867,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:d2a535662522728348ecbfc35b114fa4d4ec74c797735a3126d6d53e279fa452",
    "changeType": "UPSERT",
    "aspectName": "querySubjects",
    "aspect": {
        "json": {
            "subjects": [
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,snowflake.db.sessions,PROD),user_id)"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989867,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:d2a535662522728348ecbfc35b114fa4d4ec74c797735a3126d6d53e279fa452",
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
        "json": {
            "platform": "urn:li:dataPlatform:snowflake"
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989868,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:faa1ad49182bace324e5bebef7f86a2e1e1c7fd6726b9180840109f46bd70a51",
    "changeType": "UPSERT",
    "aspectName": "queryProperties",
    "aspect": {
        "json": {
            "customProperties": {},
            "statement": {
                "value": "SELECT * FROM temp_user_sessions WHERE session_count > 5",
                "language": "SQL"
            },
            "source": "SYSTEM",
            "created": {
                "time": 1609459380000,
                "actor": "urn:li:corpuser:analyst"
            },
            "lastModified": {
                "time": 1609459380000,
                "actor": "urn:li:corpuser:analyst"
            }
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989868,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:faa1ad49182bace324e5bebef7f86a2e1e1c7fd6726b9180840109f46bd70a51",
    "changeType": "UPSERT",
    "aspectName": "querySubjects",
    "aspect": {
        "json": {
            "subjects": [
                {
                    "entity": "urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD),session_count)"
                },
                {
                    "entity": "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:snowflake,temp_user_sessions,PROD),user_id)"
                }
            ]
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989868,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
},
{
    "entityType": "query",
    "entityUrn": "urn:li:query:faa1ad49182bace324e5bebef7f86a2e1e1c7fd6726b9180840109f46bd70a51",
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
        "json": {
            "platform": "urn:li:dataPlatform:snowflake"
        }
    },
    "systemMetadata": {
        "lastObserved": 1757897989869,
        "runId": "sql-queries-2025_09_14-20_59_49-j4q8hq",
        "lastRunId": "no-run-id-provided"
    }
}
]